src/db.c

```diff
@@ -1604,25 +1604,143 @@ int *getKeysPrepareResult(getKeysResult *result, int numkeys) {
     if (numkeys > result->size) {
         if (result->keys != result->keysbuf) {
             /* We're not using a static buffer, just (re)alloc */
-            result->keys = zrealloc(result->keys, numkeys * sizeof(int));
+            result->keys = zrealloc(result->keys, numkeys * sizeof(keyReference));
         } else {
             /* We are using a static buffer, copy its contents */
-            result->keys = zmalloc(numkeys * sizeof(int));
+            result->keys = zmalloc(numkeys * sizeof(keyReference));
             if (result->numkeys)
-                memcpy(result->keys, result->keysbuf, result->numkeys * sizeof(int));
+                memcpy(result->keys, result->keysbuf, result->numkeys * sizeof(keyReference));
         }
         result->size = numkeys;
     }
 
     return result->keys;
 }
 
+/* Fetch the keys based of the provided key specs. Returns 0 if the keyspec definition is
+ * incomplete. */
+int getKeysUsingKeySpecs(struct redisCommand *cmd, robj **argv, int argc, getKeysResult *result) {
+    int j, i, k = 0, last, first, step;
+    keyReference *keys;
+
+    for (j = 0; j < cmd->key_specs_num; j++) {
+        keySpec *spec = cmd->key_specs+j;
+        serverAssert(spec->begin_search_type!=KSPEC_BS_INVALID);
+        if (spec->flags & CMD_KEY_INCOMPLETE) {
+            result->numkeys = 0;
+            return 0;
+        }
```

Comment:
Based on other communication, I think if there is an incomplete flag we want to abort and go to the authority which is the function call.

Reply:
i'm thinking of this function in a more generic term.
it's purpose is to find all the keys and return an indication of the flags, and it can someday be used elsewhere.
so i'd prefer to conclude the search and find all the keys that could be found.
if we want an early abort optimization, maybe that could be an input flag?
if not, let's at least make sure the function's top comment mentions that the search is aborted early as soon as this flag is found.

Reply:
I don't see the need to pre-optimization for that case, since you should call the other function to fetch the keys.

Reply:
by unneeded pre-optimization you mean to about the search early as soon as the incomplete flag is found (optimization)?
or keep searching to return a list as full as possible?
i.e. which one of these is unneeded?

i consider this a reference implementation of how to extract keys from key-specs, i don't know how it'll be used in the future (inside redis), and would even rather think that people will copy that code from redis to use in other projects.
thus, i'd rather let it keep the search, and not abort here.

Reply:
If we want this function be a complete reference implementation for fetching keys, we may want to call the get_keys_proc() here instead of aborting.

Reply:
Then let's provide clients with an implementation that get's one key for cluster mode support. I don't think this code should be a model for an unknown number of client implementations. This code could be much simpler if it's just searching for a single key.

Reply:
i rather not maintain two implementations.
i think that with minimal changes to this one, it can serve all purposes.
i.e. add some input flags like:
1. exit ASAP as soon as an INCOMPLETE is found.
2. exit ASAP as soon as one key is found.

actually, for the first purpose, we don't even need to scan the arguments, we can just add a short 3 line function that runs on the key-specs and returns an & mask of all flags.

and we can even ignore the second purpose, users can easily edit the code and add a return as soon as one key is found.

so what i'd prefer is:
1. add a new function to extract all key-spec flags and avoid calling this one at all on commands that have INCOMPLETE.
2. remove the handling of the INCOMPLETE flag in this function, and leave the code able to extract all possible key names it can.

Reply:
Fair enough, I'll update this. It's not something I care that much about.

Reply:
@oranagra I didn't quite implement your suggestion, but I did something similar. Let me know what you think:

This function has two optional search flags: GET_KEYSPEC_IGNORE_CHANNELS and GET_KEYSPEC_RETURN_PARTIAL. Channels is straightforward, but partial is used both to check for incomplete flags as well as allows for syntax checking. On syntax errors, the default behavior is now to return nothing, back like in the earlier revisions, but when GET_KEYSPEC_RETURN_PARTIAL is passed in it will still return all of the valid keys found.

Also note that in an earlier revision there was clamping, and I decided that potentially return false keys was not a good behavior. (If a client was trying to route for cluster mode, it might route to the wrong instance). So now it just continues.

Reply:
i'll review it soon. but first 2 questions:
1. so it means that redis uses the PARTIAL flag (to abort on the first sign of INCOMPLETE), and if cluster clients will copy that code, they'll use partial too in order to get just one key? doesn't it mean they may get 0 keys (in case for SORT the first key-spec has the INCOMPLETE flag)?
2. don't you think that for redis, for the purpose of deciding if we wanna use the getkeys_proc or the key-specs, it's easier and faster to just add a 3 line function that iterates on all key-specs and returns a mask of all flags? (without caring about the args)